library/std/src/sys/pal/wasip2/cabi_realloc.rs

```diff
@@ -0,0 +1,65 @@
+//! This module contains a canonical definition of the `cabi_realloc` function
+//! for the component model.
+//!
+//! The component model's canonical ABI for representing datatypes in memory
+//! makes use of this function when transferring lists and strings, for example.
+//! This function behaves like C's `realloc` but also takes alignment into
+//! account.
+//!
+//! Components are notably not required to export this function, but nearly
+//! all components end up doing so currently. This definition in the standard
+//! library removes the need for all compilations to define this themselves.
+//!
+//! More information about the canonical ABI can be found at
+//! <https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md>
+//!
+//! Note that the name of this function is not standardized in the canonical ABI
+//! at this time. Instead it's a convention of the "componentization process"
+//! where a core wasm module is converted to a component to use this name.
+//! Additionally this is not the only possible definition of this function, so
+//! this is defined as a "weak" symbol. This means that other definitions are
+//! allowed to overwrite it if they are present in a compilation.
+
+use crate::alloc::{self, Layout};
+use crate::ptr;
+
+#[used]
+static FORCE_CODEGEN_OF_CABI_REALLOC: unsafe extern "C" fn(
+    *mut u8,
+    usize,
+    usize,
+    usize,
+) -> *mut u8 = cabi_realloc;
+
+#[linkage = "weak"]
+#[no_mangle]
+pub unsafe extern "C" fn cabi_realloc(
```

Comment:
I know that liballoc generally does not have OS-specific code, but would it make sense to make an exception for `cabi_realloc`?

`cabi_realloc` doesn't do any "syscalls" or otherwise depend on std or `libc`. It does call `abort_internal` which calls `libc::abort`, but that's [just a `__builtin_trap` call](https://github.com/WebAssembly/wasi-libc/blob/main/libc-bottom-half/sources/abort.c#L3), so we could do that inline instead of calling libc.

Reply:
Oh this is in libstd, not liballoc, but do you think it'd be best to go straight to `unreachable` here?

Reply:
Is there a reason to move this to liballoc though? I'd rather not try to wade too much into adding platform-specific bits to liballoc when there aren't really any there currently.

Reply:
Is there much reason for wasi code to be no_std though? Is there a downside to using std for this?

Reply:
I`ve found `no_std` to be useful when creating components that strictly conform to a given world. In a `std` build, something like the `assert!` macro pulls in code for environment variables and stderr, but in `no_std` it doesn't.

Reply:
Or, wait, `panic = abort` fixes that.

Reply:
Ok that makes sense, but that may also want to be something where a feature is added to libstd to not print on failure or something similar to avoid extra host imports perhaps?

Reply:
Would a feature in libstd would depend on `-Z build-std` to use, and thus Rust nightly? I can see why it's desirable to head in that direction, though it is less convenient in the present.

Reply:
Correct yeah, it would require rebuilding there. I'm basically hesitant to spread this out too much because this is "The Spot" for platform-specific code. I don't want to add any maintenance burden for liballoc having to think about platform-specific things and instead keep it all here in libstd.

I'll admit though that I'm not personally very motivated by `no_std` stuff in wasm since the standard library is there to use, so that's also coloring my opinion here. Working around `assert!` and and other size-related concerns like that doesn't seem like great motivation to me for introducing the concept of platform-specific-ness in an otherwise platform-independent crate.

Reply:
It looks like `panic_immediate_abort` is the thing we need here. I can confirm that `-Z build-std=std,panic_abort -Z build-std-features=panic_immediate_abort` builds a Wasm component that does `assert!` with no extra imports.

It's inconvenient that this depends on Nightly Rust, but we can revisit this later if needed.